proxy.go

```diff
@@ -174,13 +174,22 @@ func executeWithRetry(
 	monitorDuration func(float64),
 	monitorRetryRequestInc func(prometheus.Labels),
 ) (float64, error) {
-	// num of replicas should > 1,
-	// when the host is unavailable,
-	// it could make sure there might be another set of replicas contains all the data of clickhouse
 	startTime := time.Now()
 	var since float64
-	numReplicas := len(s.host.replica.cluster.replicas)
-	if numReplicas > 1 && retryNum <= numReplicas {
+
+	// Check all active hosts
+	numHosts := 0
```

Comment:
you don't even need the s.cluster.replicas
you can simplify the logic of the full function and avoid the (if numHosts > 1 {...} else {..} )
the idea in pseudo code is to do something like

`shouldRetry = false
numretry = 0
do {
  res, err := run_query(host)
  if err == 502 && numretry <maxRetry {
    shouldRetry = true
   [all the logic to penalize then change the host]
  } else {
    shouldRetry = false
  }
}
`

Reply:
Done